TF implementation of Inception Time - optimized for each task using Grid Search

Bsingstad opened the pull request:
In this pull request, we propose a new implementation of Inception Time (Tensorflow) which parameters are optimized using grid search. This work will be further explained in a paper published in ReScience C journal :   http://rescience.github.io/

Bsingstad committed:
push NBs

Bsingstad committed:
updated notebook

Bsingstad committed:
Do changes in order to be able to run on Google Colab

Bsingstad committed:
Update PTB_XL_experiments.ipynb

Bsingstad committed:
set output folder and change NB

Bsingstad committed:
override hardcoded folders

Bsingstad committed:
Make it possible to store results in google drive

Bsingstad committed:
Update reproduce_results.py

Bsingstad committed:
main take folder as arg and add correct args to evaluate in order to reprod results

Bsingstad committed:
Add support for using arguments when running file

Bsingstad committed:
Update reproduce_results.py

Bsingstad committed:
Update scp_experiment.py

add noise to test data to assess robustness

Bsingstad committed:
Add my model

Bsingstad committed:
small changes to noise induction

Bsingstad committed:
fix

Bsingstad committed:
Update your_configs.py

Bsingstad committed:
change config

Bsingstad committed:
FIX: found a change that had to be done in order to run my_model

epoch = 2 for my model in order to test.

Bsingstad committed:
Update reproduce_results.py

Add right config name

Bsingstad committed:
Add IF-statement to control if noise should be added or not

Bsingstad committed:
change predict statment and set noise as False by default

Bsingstad committed:
Update to 30 epochs

Bsingstad committed:
Add my finetuning NB

Bsingstad committed:
Update your_model.py

Bsingstad committed:
fix errors

Bsingstad committed:
Add grid search files

Bsingstad committed:
Update My_Finetuning.ipynb

Bsingstad committed:
Create My_Finetuning2.ipynb

Bsingstad committed:
Update My_Finetuning2.ipynb

Save results from gridsearch to Google Drive (save each gridsearch round to ensure that results gets stored)

Bsingstad committed:
Add files via upload

Bsingstad committed:
Update My_Finetuning170223.ipynb

Bsingstad committed:
Add files via upload

Bsingstad committed:
Update model params after grid search

Bsingstad committed:
add tf addons

Bsingstad committed:
add results from optimized Inception model

Bsingstad committed:
update models to run

Bsingstad committed:
Delete make_gridsearch_file.ipynb

Bsingstad committed:
Delete gridsearch_params.csv

Bsingstad committed:
Delete final_results.txt

Bsingstad committed:
Delete PTB_XL_experiments.ipynb

Bsingstad committed:
Delete My_Finetuning170223.ipynb

Bsingstad committed:
Delete My_Finetuning2.ipynb

Bsingstad committed:
Delete Gridsearch280323.ipynb

Bsingstad committed:
Delete My_Finetuning.ipynb

Bsingstad committed:
delete GS folder

Bsingstad committed:
Merge branch 'repo_for_PR' of https://github.com/Bsingstad/ecg_ptbxl_benchmarking into repo_for_PR

helme commented:
Hi @Bsingstad , first of all I have to excuse myself for the very (very) late reply on this PR! big sorry for that!

Overall, your PR looks great, however I need to request some changes. In particular:
- copying and renaming the template files (`your_config.py `and `your_model.py`) to a new files with better names, i.e. `tf_inception_time_config.py `and `inception_time.py`
- we want to keep `code/reproduce_results.py` as is in order to allow others to reproduce our results. Instead we suggest to create a new file called `code/reproduce_inception_time_results.py`
- in addition we observed, that your current version of `code/reproduce_results.py` would train all your models for each of the tasks, i.e. a model designed for `form` (`conf_tf_inception_form`) is also trained on other tasks. Is this intended, or do you prefer specific models for specific tasks? 
- an lastly, we prefer to keep the case `elif modeltype == "YOUR_MODEL_TYPE"` as a template, instead we would suggest to just add an additional case for your models here

other than that, your PR looks great, I'm curious to do pull request once the issues are resolved. What do you think about our suggestions?

Best, 
Patrick

Bsingstad commented:
Hi @helme

Thank you for your suggestions. I will address them as soon as I can.

Best regards
Bjørn-Jostein

Bsingstad committed:
PR Update

Here I have updated the pull request according to the requirments from the repository owner

Bsingstad committed:
add template to scp_experiment.py

Bsingstad commented:
Hi @helme 

Sorry for taking so long time.

I have now updated the pull request according to your requirements:

1.  Good point! I have renamed the files containing the code for our proposed model to `inception_time.py` and configs `tf_inception_time_config.py`. We also added the original template files `your_model.py` and `your_configs.py` to our PR.
2.  I have renamed `code/reproduce_results.py` to  `code/reproduce_inception_time_results.py` in my PR
3. Actually this was an unintended behavior. In our replication paper we only report the specific model for each specific task, like this:

| **Model**                       | **All**        | **Diagnostic**  | **Subdiagnostic**  | **Superdiagnostic**  | **Form**      | **Rhythm**      |
|---------------------------------|----------------|-----------------|--------------------|----------------------|---------------|-----------------|
| **Inception Time (all)**        | 0.926(08)      |                 |                    |                      |               |                 |
| **Inception Time (diagnostic)** |                | 0.929(09)       |                    |                      |               |                 |
| **Inception Time (subdiagnostic)** |              |                 | 0.927(08)          |                      |               |                 |
| **Inception Time (superdiagnostic)** |            |                 |                    | 0.922(06)           |               |                 |
| **Inception Time (form)**       |                |                 |                    |                      | 0.840(11)     |                 |
| **Inception Time (rhythm)**     |                |                 |                    |                      |               | 0.923(32)       |

but we didn't find a way to only train the specific models without doing major changes in your base code.
4. Of course! I have added `elif modeltype == "YOUR_MODEL_TYPE"` to `scp_experiment.py` and added an additional case for my model:  `elif modeltype == "inception_time_model":`

helme commented on code/utils/utils.py:
```diff
@@ -114,7 +114,7 @@ def apply_thresholds(preds, thresholds):
 
 def load_dataset(path, sampling_rate, release=False):
-    if path.split('/')[-2] == 'ptbxl':
+    if path.split('/')[-2] == 'ptb-xl-a-large-publicly-available-electrocardiography-dataset-1.0.1':
```
in [get_datasets.sh line 10](https://github.com/helme/ecg_ptbxl_benchmarking/blob/516740dd2964d67c213ab6df9eba5d50b2245d00/get_datasets.sh#L10) the folder is renamed to `ptbxl`. So, we would like to revert this change.

helme commented on code/utils/utils.py:
```diff
@@ -335,7 +335,7 @@ def apply_standardizer(X, ss):
 # DOCUMENTATION STUFF
 
-def generate_ptbxl_summary_table(selection=None, folder='../output/'):
+def generate_ptbxl_summary_table(selection=None, folder: str="./your/path/to/ptbxl/"):
```
since this folder `./your/path/to/ptbxl/` does not exist by default, we would like to keep `../output/`.

helme commented on code/utils/utils.py:
```diff
@@ -405,7 +405,7 @@ def generate_ptbxl_summary_table(selection=None, folder='../output/'):
     print(md_source)
 
-def ICBEBE_table(selection=None, folder='../output/'):
+def ICBEBE_table(selection=None, folder:str="./your/path/to/icbeb/"):
```
the same for this argument.

helme commented on code/experiments/scp_experiment.py:
```diff
@@ -113,6 +124,8 @@ def perform(self):
             model.predict(self.X_val).dump(mpath+'y_val_pred.npy')
+            plt.plot(self.X_test[0,:,0])
+            plt.show()
             model.predict(self.X_test).dump(mpath+'y_test_pred.npy')
```
this should be removed, since it's probably a debugging statement.

helme commented on code/experiments/scp_experiment.py:
```diff
@@ -5,6 +5,7 @@
 import numpy as np
 import multiprocessing
 from itertools import repeat
+import matplotlib.pyplot as plt
```
this is not needed, if the last debugging output is removed

helme commented on code/experiments/scp_experiment.py:
```diff
@@ -23,6 +24,8 @@ def __init__(self, experiment_name, task, datafolder, outputfolder, models, samp
         self.outputfolder = outputfolder
         self.datafolder = datafolder
         self.sampling_frequency = sampling_frequency
+        self.noise_mean = 0
+        self.noise_std_scale = 0.1
```
this is fine.

helme commented on code/reproduce_results.py:
this file should have been kept as is and copied and renamed to [code/reproduce_inception_time_results.py](https://github.com/helme/ecg_ptbxl_benchmarking/pull/23/files#diff-379b72794183c23f6d8e0e90fec8196698957fa6da11f13d1436486e1e2ae41e)

helme replied:
in [this branch](https://github.com/helme/ecg_ptbxl_benchmarking/blob/5c6e02c167ff9dd21cf2b965e8cd9dcc1de127c6/code/reproduce_inception_time_results.py)  I proposed a workaround to your problem, does this work for you? If yes, integrate it to this branch please and commit to the PR.

helme commented on code/reproduce_inception_time_results.py:
in [this branch I proposed a workaround to your problem](https://github.com/helme/ecg_ptbxl_benchmarking/blob/5c6e02c167ff9dd21cf2b965e8cd9dcc1de127c6/code/reproduce_inception_time_results.py)

helme committed:
Update scp_experiment.py

removed matplotlib and debugging stuff

helme commented on code/models/inception_time.py:
nice.

helme committed:
Update utils.py

did the changes according to my comments

helme reviewed (changes requested):
Hi, first of all, thank your very much for this PR, looks promising :)

However, I still request some changes such that the reproduction of the results is still possible. Could you do this?

helme commented:
Hi @Bsingstad,
I hope my change requests for this pull request are somehow understandable?

Bsingstad reviewed (commented):
I am satisfied with your change requests

Bsingstad committed:
Requested changes

Bsingstad commented:
I have now committed the changes based on your last review

Bsingstad committed:
change file path names
